### PR TITLE
[RFC] os_unix_defs: Consistently use '~' over '$HOME'

### PR DESCRIPTION
**Pull request by ghost**
From #889: `I do that because nvim --version looked ugly.`

I don't think this will be an issue given that `os_unix_defs.h` already uses both `$HOME` and `~`, but I've only tried this on Linux with a few different shells. If anyone could try it on Windows that'd be appreciated.

Also it's worth mentioning that I'm not dead set on using `$HOME`; `~` is fine, as long as it's consistent.

### CODE REVIEW

**Comment by justinmk** [comment]
It really depends on what these are being fed into. By the way `$HOME` often does not exist on Windows, so I suspect `~` might be better since n/vim can be expected to do the right thing with that.

**Comment by Pyrohh** [comment]
> Whereas ~ is always handled like a hot potato by Windows gvim, so it can never be "taken literally".

What do you mean by that? Isn't it a good thing that `~` is never literally interpreted, because it almost always expands to a directory not named `~`.

I must be missing something here...

**Comment by justinmk** [comment]
> What do you mean by that? Isn't it a good thing that ~ is never literally interpreted

Yes, exactly. Since we can count on Vim doing the right thing with `~`, I'm speculating that we can just use that across the board instead of inspecting every code path to make sure `$HOME` falls back appropriately. I'm pretty sure it doesn't matter, but isn't `~`  nicer? :)

**Comment by Pyrohh** [comment]
> Yes, exactly. Since we can count on Vim doing the right thing with ~, I'm speculating that we can just use that across the board instead of inspecting every code path to make sure $HOME falls back appropriately. I'm pretty sure it doesn't matter, but isn't ~ nicer? :)

Sounds good to me, and yes `~` does look nicer :)

**Comment by Pyrohh** [comment]
Repushed with the `~` changes, ping @justinmk

**Comment by kopischke** [comment]
Shouldn’t the commit message and PR title be changed to effect the reversal of the original intention?

**Comment by Pyrohh** [comment]
Thanks for mentioning that @kopischke, it has been done.

**Comment by justinmk** [comment]
Can you do a quick test that it doesn't break if HOME is unset?

**Comment by Pyrohh** [comment]
I've tried mksh, bash, zsh, and sh (which on my system is just a symlink to bash), and none of them allow me to unset HOME. They all seem to work with `HOME=""`, but I don't think that's what you're asking.

File: src/nvim/os_unix_defs.h [hunk 1]
```diff
@@ -145,7 +145,7 @@
 
 
 #ifndef USR_GVIMRC_FILE
-#  define USR_GVIMRC_FILE "$HOME/.ngvimrc"
+#  define USR_GVIMRC_FILE "~/.ngvimrc"
 #endif
```
Comment (justinmk) [inline src/nvim/os_unix_defs.h]: We can delete USR_GVIMRC_FILE , if you want

File: src/nvim/os_unix_defs.h [hunk 2]
```diff
@@ -145,7 +145,7 @@
...
 
 #ifndef USR_GVIMRC_FILE2
```
Comment (justinmk) [inline src/nvim/os_unix_defs.h]: this too is not used

**Comment by justinmk** [comment]
`unset HOME` works for me in bash/unbuntu. With an older build of nvim (before this PR):

```
unset HOME
echo $HOME
<nothing>
nvim
<quit>
E138: Can't write viminfo file $HOME/.nviminfo!
``` 

With this PR:

```
unset HOME
echo $HOME
<nothing>
nvim
<quit>
E138: Can't write viminfo file ~/.nviminfo!
```

In both cases, my `~/.nvimrc` is not found (equivalent to `nvim -u NONE`). So, there is no regression by this PR. It might be click to fall back to `echo ~` if the shell supports it, but out of scope for now I guess.

**Comment by Pyrohh** [comment]
I was doing `unset HOME`, forgot the `$`.

> It might be slick to fall back to echo ~ if the shell supports it, but out of scope for now I guess

Ehhhhhhhh... I'm not so sure about that. For all we know the user is doing it on purpose, like `nvim -u NONE`. As you say it's out of scope, but I wasn't sure if you meant for this PR, or to be done in the future.

http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06_01

File: src/nvim/os_unix_defs.h (hunk 2 above)
Comment (Pyrohh) [inline src/nvim/os_unix_defs.h]: Can SYS_GVIMRC_FILE be removed as well? Given that nvim will defer to external programs for GUIs, I don't think it's needed.

Comment (justinmk) [inline src/nvim/os_unix_defs.h]: yes.

**Comment by justinmk** [comment]
:+1:

**Comment by Pyrohh** [comment]
reeeeeeeeeeeeeepushed

**Comment by justinmk** [comment]
merged, thanks. I reverted the main.c changes because they did not make sense, given the implementation of `init_homedir()`.

**Comment by Pyrohh** [comment]
> I reverted the main.c changes because they did not make sense, given the implementation of init_homedir().

Oops, didn't think that one through. Thanks.

**Pull request closed without merging**